Don't detect MSYS-translated constructs as ordinary files

From: J.M. Eubank <john@thesnappy.net>

From
   <https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-binutils/0002-check-for-unusual-file-harder.patch>

Configure scripts tend to do things like this:
nm -B /dev/null
which, on Windows, turn into:
nm -B nul
and nul here acts like a normal file with 0 size as far as stat() is concerned.
Check harder for the file in question being unusual (isatty() does see that).
And since configure script expects /dev/nul, do report it that way.
---
 binutils/bucomm.c  |   20 ++++++++++++++++++--
 binutils/elfedit.c |   15 ++++++++++++++-
 2 files changed, 32 insertions(+), 3 deletions(-)

diff --git a/binutils/bucomm.c b/binutils/bucomm.c
index 9e6a0284..0692a9c0 100644
--- a/binutils/bucomm.c
+++ b/binutils/bucomm.c
@@ -608,6 +608,15 @@ get_file_size (const char * file_name)
   if (file_name == NULL)
     return (off_t) -1;
 
+  int f, t;
+  t = -1;
+  f = open (file_name, O_RDONLY | O_BINARY);
+  if (f != 0)
+    {
+      t = isatty (f);
+      close (f);
+    }
+
   if (stat (file_name, &statbuf) < 0)
     {
       if (errno == ENOENT)
@@ -618,8 +627,15 @@ get_file_size (const char * file_name)
     }
   else if (S_ISDIR (statbuf.st_mode))
     non_fatal (_("Warning: '%s' is a directory"), file_name);
-  else if (! S_ISREG (statbuf.st_mode))
-    non_fatal (_("Warning: '%s' is not an ordinary file"), file_name);
+  else if (! S_ISREG (statbuf.st_mode) || t > 0)
+    {
+#ifdef _WIN32
+      /* libtool passes /dev/null and checks for /dev/null in the output */
+      if (stricmp (file_name, "nul") == 0)
+        file_name = "/dev/null";
+#endif
+      non_fatal (_("Warning: '%s' is not an ordinary file"), file_name);
+    }
   else if (statbuf.st_size < 0)
     non_fatal (_("Warning: '%s' has negative size, probably it is too large"),
                file_name);
diff --git a/binutils/elfedit.c b/binutils/elfedit.c
index 7c475c90..0826e753 100644
--- a/binutils/elfedit.c
+++ b/binutils/elfedit.c
@@ -691,6 +691,14 @@ static int
 check_file (const char *file_name, struct stat *statbuf_p)
 {
   struct stat statbuf;
+  int f, t;
+  t = -1;
+  f = open (file_name, O_RDONLY | O_BINARY);
+  if (f != 0)
+    {
+      t = isatty (f);
+      close (f);
+    }
 
   if (statbuf_p == NULL)
     statbuf_p = &statbuf;
@@ -705,8 +713,13 @@ check_file (const char *file_name, struct stat *statbuf_p)
       return 1;
     }
 
-  if (! S_ISREG (statbuf_p->st_mode))
+  if (! S_ISREG (statbuf_p->st_mode) || t > 0)
     {
+#ifdef _WIN32
+      /* libtool passes /dev/null and checks for /dev/null in the output */
+      if (stricmp (file_name, "nul") == 0)
+        file_name = "/dev/null";
+#endif
       error (_("'%s' is not an ordinary file\n"), file_name);
       return 1;
     }
